feat: add Badge component

langz opened the pull request:
This PR adds the Badge component.
You can see initial Figma-sketches for the component here https://www.figma.com/file/cdtwQD8IJ7pTeE45U148r1/Eufemia---Web?node-id=17917%3A43

Here is the latest Figma-sketches, with a updated design, https://www.figma.com/file/cdtwQD8IJ7pTeE45U148r1/Eufemia---Web?node-id=26322%3A25223

Take a look at it here: https://eufemia-featbadgecomponent.gtsb.io/uilib/components/badge

thaytharma commented on packages/dnb-eufemia/src/components/badge/style/_badge.scss:
```diff
@@ -0,0 +1,155 @@
+  --badge-line-height--x-large: var(--line-height-basis);
+  --badge-width--small: 0.5rem;
+  --badge-height--small: 0.5rem;
+  --badge-border-radius--small: 0.25rem;
```
Just a suggestion:

For the border-radiuses below: It seems like they are half the size of their `size` (ref my `--badge--small`suggestion).

Maybe we can use the `calc`-function with the `size`-variable here instead of hard-coding its sizes? 🤔
something like `calc(#{--badge--small} / 2)`

thaytharma reviewed (commented):
Great great work! @langz 😍

langz commented:
This feature is not in a hurry as of now. Postponing development of this.

langz commented on packages/dnb-design-system-portal/src/docs/uilib/components/badge/info.md:
```diff
@@ -0,0 +1,7 @@
+
+## Description
+
+Badge
```
Added the text suggestion 👍

langz commented on packages/dnb-eufemia/src/components/badge/__tests__/Badge.test.tsx:
```diff
@@ -0,0 +1,123 @@
+    )
+  })
+
+  it('renders class "has-text" when string or number as content', () => {
```
Split this into two tests

langz commented on packages/dnb-eufemia/src/components/badge/style/_badge.scss:
```diff
@@ -0,0 +1,155 @@
+  --badge-line-height--large: var(--line-height-x-small);
+  --badge-line-height--x-large: var(--line-height-basis);
+  --badge-width--small: 0.5rem;
+  --badge-height--small: 0.5rem;
```
Sure sure, I've just tried following the same pattern as else where, like for button.
https://github.com/dnbexperience/eufemia/blob/4f6ce0fb88cd69e307c5bbd2b06b359b3c2239c5/packages/dnb-eufemia/src/components/button/style/_button.scss#L9-L16

langz commented on packages/dnb-eufemia/src/components/badge/__tests__/Badge.test.tsx:
```diff
@@ -0,0 +1,123 @@
+    expect(within(badgeRoot).queryByTestId('confetti-icon')).not.toBeNull()
+  })
+  describe('default values', () => {
+    it('has variant primary as default', () => {
```
Done

langz commented:
I was thinking to start looking into this again, but I now see that the design is change and quite different from last time.
https://www.figma.com/file/cdtwQD8IJ7pTeE45U148r1/Eufemia---Web?node-id=26322%3A25223

thaytharma commented:
> I was thinking to start looking into this again, as @thaytharma will need this in funds, but I now see that the design is changed and looks quite different from last time.
> 
> https://www.figma.com/file/cdtwQD8IJ7pTeE45U148r1/Eufemia---Web?node-id=26322%3A25223
> 
> Does this design look like what you need @thaytharma? Or is what you need more like https://eufemia-featbadgecomponent.gtsb.io/uilib/components/badge ?
> 
> EDIT: I've added a few changes now @thaytharma, to hopefully accommodate to your needs. Do you have time to give it a quick spin? :) https://eufemia-featbadgecomponent.gtsb.io/uilib/components/badge

This is exactly what we want,
Thanks @langz 🙏🏼

thaytharma reviewed (approved):
Well done @langz!! 😍

langz commented:
@tujoworker, should I remove the [styling of badge](https://github.com/dnbexperience/eufemia/blob/d5e0126a71fbdbd903d0ccb4d2870f90976e3371/packages/dnb-eufemia/src/components/badge/style/_badge.scss#L199-L210) when children is payment-card? [Visual example](https://eufemia-featbadgecomponent.gtsb.io/uilib/components/badge#alternatives).

To cater for the [red colored notification badge](https://www.figma.com/file/cdtwQD8IJ7pTeE45U148r1/Eufemia---Web?node-id=26322%3A25224), do you think it's okay to add a variant called `signal`? [Visual example](https://eufemia-featbadgecomponent.gtsb.io/uilib/components/badge#variant-signal)?

tujoworker commented:
I check with the designers about the additional variants and sizes. I fear that they do want it more restricted 🥶 in terms of flexibility.

langz commented:
> I check with the designers about the additional variants and sizes. I fear that they do want it more restricted 🥶 in terms of flexibility.

Since last time this was reviewed, I've now:
- Removed prop `size`, and the possibility for having different sizes.
- Removed prop `variant`, so the color of the badge is based on the "state"(information or notification).
- information state is now the default state.
- Removed obsolete tests, demos, etc.

langz commented on packages/dnb-eufemia/src/components/badge/Badge.tsx:
```diff
@@ -0,0 +1,164 @@
+
+    if (stateIsNotification && !contentIsNum) {
+      warn(
+        `Type of content should be a number: A notification badge is best suited to display content of type number.`
```
@tujoworker, do you have any other ideas on how I could sort this one out?
I would like for the type of content of a badge with state notification to be a number.

tujoworker commented on packages/dnb-design-system-portal/src/docs/uilib/components/badge.md:
```diff
@@ -0,0 +1,12 @@
+---
+title: 'Badge'
+description: 'The badge component.'
```
… to let the user focus on new or unread  content or notifications. Something that describes is a bit better?

langz replied:
Thanks, will add this 🙏

tujoworker commented on packages/dnb-eufemia/src/components/badge/Badge.tsx:
```diff
@@ -0,0 +1,164 @@
+      >
+        {label && (
+          <span data-testid="badge-label" className="dnb-sr-only">
+            {label}
```
When a label is given, we need a white space, else a screen reader reads it like this: New9+

tujoworker commented on packages/dnb-eufemia/src/components/badge/__tests__/Badge.test.tsx:
```diff
@@ -0,0 +1,147 @@
+  })
+
+  it('renders 9+ when content is a number with value greater than 9', () => {
+    const number = 100
```
Could we test with 10, else the code could have a failure (false/positive) up to 99

tujoworker commented on packages/dnb-eufemia/src/components/badge/__tests__/Badge.test.tsx:
```diff
@@ -0,0 +1,147 @@
+  it('renders the label as string', () => {
+    const label = 'Money:'
+    render(<Badge label={label} content={100} />)
+    expect(screen.queryByTestId('badge-label')).not.toBeNull()
```
Not really needed `not.toBeNull `. Interesting would also be to query the parent element and use `element.textContent` – this way you get everything inside the badge like a “screen reader”.

langz replied:
Thanks, I've improved upon this now 🙏

tujoworker commented on packages/dnb-eufemia/src/components/badge/__tests__/Badge.test.tsx:
```diff
@@ -0,0 +1,147 @@
+
+describe('Badge aria', () => {
+  it('should validate', async () => {
+    const Component = render(<Badge content="1" />)
```
Because we spy the console log before, we do not get a warn message. A label is required here.

tujoworker replied:
Yeah, but I would argue that this is a responsibility devs takes serious and it does not violent regulations or UX as it should/will be catchend before prod. While a11y warnings we do because of UX and regulations.

langz replied:
Thanks, I am not 100% sure what the outcome of this is, so I am now trying to formulate myself in great detail 😂 

In the component as of right now:
- The `label` property is optional.
- We only "warn" the dev that he should add a `label` when the content is of type `number`, and label is not provided.

Based on this discussion, should we:
- Make the `label` property required?
- Always "warn" the dev when label is not provided?
- Just add a label property to the code line("should validate"-test) we are commenting on?

tujoworker replied:
You did solve the label warning 100% 👌 so I guess I mixed things up with the other warning we had 😇 sorry for that 👼

langz replied:
Hehe, thanks for looking into this 💯

tujoworker reviewed (approved):
Veeery good work 🙌 
I got feedback that Mats and Matthias will first in august make the decision if PaymentCard should have a Badge on the card or on the heading. I would remove it for now. But it’s up to you to decide 😉

langz commented:
> Veeery good work 🙌 I got feedback that Mats and Matthias will first in august make the decision if PaymentCard should have a Badge on the card or on the heading. I would remove it for now. But it’s up to you to decide 😉

Thanks, I'll remove it for now then 🙏

tujoworker commented:
:tada: This PR is included in version 9.28.0 :tada:

The release is available on:
- [npm package (@latest dist-tag)](https://www.npmjs.com/package/@dnb/eufemia/v/9.28.0)
- [GitHub release](https://github.com/dnbexperience/eufemia/releases/tag/v9.28.0)

Your **[semantic-release](https://github.com/semantic-release/semantic-release)** bot :package::rocket:
